### ISSUE TITLE
Only half of original entropy for the third shard

### ISSUE THREAD

**Issue by EDmitry** [MJL85/seedpart#1]
The third shard will have only half of the original entropy, in other words, if the original seed was 12 words long, it's like having 6 words instead (bad). I wonder if there is a better way to generate the third shard.

**Comment by nifri** [comment]
@EDmitry, could you please elaborate why the entropy of the third shard would only have half the original entropy?

I was wondering about the information theoretic security of the third shard myself. But I'm lacking the background to judge for myself.

**Comment by EDmitry** [comment]
Let's say you only have two words in the seed: {a, b}. Now we create another random entropy, also with two words: {c, d}, this becomes our first shard.

The second shard is a result of xor'ing seed with entropy:

e = a xor c
f = b xor d

The third one is derived using the following:

g = c xor f
h = d xor e

Now, let's say an attacker got access to the third shard {g, h}

g xor h = c xor f xor d xor e = <substitue e and f from above> =  b xor a

This means that an attacker could only brute force one word from this particular seed, for example, a, and b would be derived from it explicitly using the xor operation. For seeds longer than two words this means that an attacker would only have to brute force half of all words in the original seed and can always derive the other half, using the information provided by the third shard.

In other words, the third shard discloses some information about relations between words in the original seed.

Does this make sense?

**Comment by nifri** [comment]
Now it is clear. Thanks for taking the time to explain! Your PS especially helps me to better judge for myself if I should use this or an alternative.

**Comment by EDmitry** [comment]
I'd also propose an alternative approach. Let's start with a 3 out of 3 scheme. Generate two shards randomly, and the third one would be equal to the original seed and the two of the previous shards xor'ed together (all words in order). You would need to have all three pieces to restore an original seed. Each shard would have 12 words.

Now to make it two-out-of-three, you could prepare 3 batches, each includes 2 shards:

1. A, B
2. B C
3. C A

As you can easily verify, each combo of two of these batches will give you all the three shards to restore the seed. Each shard is still 12 words, each packet is 24 words just like in the algorithm suggested in this repo.

The main benefit though is that it's extremely easy to recover the seed, and the algorithm is easy to remember. You just identify 3 unique shards out of 4 that you have from two batches and xor all words together in order. No need to remember what word lists have to be reversed and how, no need to label batches or individual shards either.
